gtsam/discrete/TableFactor.h

```diff
@@ -0,0 +1,195 @@
+  static Eigen::SparseVector<double> Convert(const std::vector<double>& table);
+
+  /** Convert table to SparseVector */
+  static Eigen::SparseVector<double> Convert(const std::string& table);
+
+ public:
+  // typedefs needed to play nice with gtsam
+  typedef TableFactor This;
+  typedef DiscreteFactor Base;  ///< Typedef to base class
+  typedef boost::shared_ptr<TableFactor> shared_ptr;
+
+  /// @name Constructors
+  /// @{
+
+  /** Default constructo for I/O */
+  TableFactor();
+
+  /** Constructor from Eigen::SparseVector */
+  TableFactor(const DiscreteKeys& keys,
+              const Eigen::SparseVector<double>& table);
+
+  /** Constructor from doubles */
+  TableFactor(const DiscreteKeys& keys, const std::vector<double>& table)
+      : TableFactor(keys, Convert(table)) {}
+
+  /** Constructor from string */
+  TableFactor(const DiscreteKeys& keys, const std::string& table)
+      : TableFactor(keys, Convert(table)) {}
+
+  /// Single-key specialization
+  template <class SOURCE>
+  TableFactor(const DiscreteKey& key, SOURCE table)
+      : TableFactor(DiscreteKeys{key}, table) {}
+
+  /// Single-key specialization, with vector of doubles.
+  TableFactor(const DiscreteKey& key, const std::vector<double>& row)
+      : TableFactor(DiscreteKeys{key}, row) {}
+
+  /// @}
+  /// @name Testable
+  /// @{
+
+  /// equals
+  bool equals(const DiscreteFactor& other, double tol = 1e-9) const override;
+
+  /// print
+  void print(
+      const std::string& s = "TableFactor\n",
+      const KeyFormatter& formatter = DefaultKeyFormatter) const override;
+
+  /// @}
+  /// @name Standard Interface
+  /// @{
+
+  /// Find value for given assignment of values to variables
+  double operator()(const DiscreteValues& values) const override;
+
+  /// Multiply with DecisionTreeFactor
+  DecisionTreeFactor operator*(const DecisionTreeFactor& f) const override;
+
+  /// Multiply with TableFactor
+  TableFactor operator*(const TableFactor& f) const;
+
+  /// Get the cardinality of key j
+  size_t cardinality(Key j) const { return cardinalities_.at(j); }
+
+  /// Generate DecisionTreeFactor from TableFactor
+  DecisionTreeFactor toDecisionTreeFactor() const override;
+
+  /// Create new factor by summing all values with the same separator values
+  shared_ptr sum(size_t nrFrontals) const;
+
+  /// Create new TableFactor where the input key is eliminated
+  TableFactor eliminate(const Key key) const;
+
+  /// Create new factor by maximizing over all values with the same separator
+  TableFactor max(size_t nrFrontals) const;
+
+  /// Create new factor by maximizing over all values with the same separator
+  TableFactor max(const Ordering& frontalKeys) const;
+
+  /// @}
+  /// @name Advanced Interface
+  /// @{
+
+  /// Generate TableFactor from DecisionTreeFactor
+  TableFactor fromDecisionTreeFactor(const DecisionTreeFactor& f) const;
+
+  /// Finds value for the key at index
+  size_t lazy_cp(Key target_key, size_t index) const;
```

Comment:
Naming convention (`lazyCartesianProduct` ?) and should this not be private?

Reply:
We can make this private but we won't be able to run unittest on this function anymore. However, since the function has been tested already I'll make it private.